src/main/java/org/folio/circulation/domain/ChangeDueDateRepository.java

```diff
@@ -0,0 +1,42 @@
+package org.folio.circulation.domain;
+
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.ServerErrorFailure;
+
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.CompletableFuture;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ChangeDueDateRepository {
+
+  private static final Logger log = LoggerFactory
+    .getLogger(MethodHandles.lookup().lookupClass());
+
+  private final LoanRepository loanRepository;
+
+  public ChangeDueDateRepository(Clients clients) {
+    loanRepository = new LoanRepository(clients);
+  }
+
+  public CompletableFuture<Result<Loan>> changeDueDate(String loanId,
+    DateTime dueDate) {
+
+    log.info("Changing due date for {} to {}", loanId, dueDate);
+    try {
+      return loanRepository.getById(loanId)
+        .thenApply(r -> r.map(loan -> loan.changeDueDate(dueDate)))
+        .thenApply(r -> r.map(loan -> {
+          loan.changeAction(LoanAction.DUE_DATE_CHANGE);
+          return loan;
+        }))
+        .thenCompose(r -> r.after(loanRepository::replaceLoan))
+        .thenApply(r -> r.map(LoanAndRelatedRecords::getLoan));
+    } catch (Exception e) {
```

Comment:
Under what circumstances might an exception be thrown?

Reply:
Has it been discussed with the POs that this operation could partially succeed, and it wouldn't be easily to determine to what degree it succeeded or failed?

Reply:
One of the main goals was providing a feedback to the client about which loans failed to change due date. Loan ids, for which the change due date failed for whatever reason, are returend to the client.